File: common/src/main/java/au/com/codeka/warworlds/common/net/PacketDecoder.java
package au.com.codeka.warworlds.common.net;

import java.io.IOException;
import java.io.InputStream;

import au.com.codeka.warworlds.common.Log;
import au.com.codeka.warworlds.common.proto.Packet;
import okio.BufferedSource;
import okio.Okio;

/**
 * Decodes a stream of {@link Packet}s, as encoded by {@PacketEncoder}.
 */
public class PacketDecoder {
  private static final Log log = new Log("PacketDecoder");

  public interface PacketHandler {
    void onPacket(PacketDecoder decoder, Packet pkt, int encodedSize);
    void onDisconnect();
  }

  private final BufferedSource source;
  private final Thread thread;

  private PacketHandler handler;

  public PacketDecoder(InputStream ins, PacketHandler handler) {
    this.source = Okio.buffer(Okio.source(ins));
    this.handler = handler;
    thread = new Thread(readRunnable);
    thread.start();
  }

  public void setPacketHandler(PacketHandler handler) {
    this.handler = handler;
  }

  private final Runnable readRunnable = new Runnable() {
    @Override
    public void run() {
      try {
        while (!source.exhausted()) {
          int size = source.readIntLe();
          int flags = source.readIntLe();
          byte[] bytes = source.readByteArray(size);
          if ((flags & PacketFlags.COMPRESSED) != 0) {
            bytes = GzipHelper.decompress(bytes);
          }
          Packet pkt = Packet.ADAPTER.decode(bytes);

          handler.onPacket(PacketDecoder.this, pkt, size);
        }
      } catch(IOException e) {
        log.warning("Error decoding packet.", e);
        handler.onDisconnect();
      }
    }
  };
}

File: common/src/main/kotlin/au/com/codeka/warworlds/common/net/PacketDecoder.kt
package au.com.codeka.warworlds.common.net

import au.com.codeka.warworlds.common.Log
import au.com.codeka.warworlds.common.net.PacketDecoder
import au.com.codeka.warworlds.common.proto.Packet
import okio.BufferedSource
import okio.buffer
import okio.source
import java.io.IOException
import java.io.InputStream

/**
 * Decodes a stream of [Packet]s, as encoded by [PacketEncoder].
 */
class PacketDecoder(ins: InputStream, private var handler: PacketHandler) {
  interface PacketHandler {
    fun onPacket(decoder: PacketDecoder, pkt: Packet, encodedSize: Int)
    fun onDisconnect()
  }

  private val source: BufferedSource = ins.source().buffer()
  private val thread: Thread

  companion object {
    private val log = Log("PacketDecoder")
  }

  fun setPacketHandler(handler: PacketHandler) {
    this.handler = handler
  }

  private val readRunnable = Runnable {
    try {
      while (!source.exhausted()) {
        val size = source.readIntLe()
        val flags = source.readIntLe()
        var bytes: ByteArray? = source.readByteArray(size.toLong())
        if (flags and PacketFlags.COMPRESSED != 0) {
          bytes = GzipHelper.decompress(bytes)
        }
        val pkt = Packet.ADAPTER.decode(bytes!!)
        handler.onPacket(this@PacketDecoder, pkt, size)
      }
    } catch (e: IOException) {
      log.warning("Error decoding packet.", e)
      handler.onDisconnect()
    }
  }

  init {
    thread = Thread(readRunnable)
    thread.start()
  }
}